Windows: Experimental: ContainerD runtime

lowenna opened the pull request:
Signed-off-by: John Howard <jhoward@microsoft.com>

Fixes #22874
Fixes #38719

See each of the commit messages for more detail on the changes.

# TL;DR 

Allows the use of containerd as an experimental runtime for containers on Windows (WCOW being both process and Hyper-V isolation, as well as LCOW).

# Detail

 - Splits libcontainerd so that both remote and local can be loaded rather than conditional compile
 - Generate a consistent OCI spec that a remote (not-in-process) runtime can escape
 - Fix a couple of bugs
 - Adds ETW logging for Windows (requires bumping logrus to v1.3.0)
 - Fixes a leak if the init proc fails to start
 - Vendoring updates for `Microsoft/hcsshim` and `Microsoft/go-winio`

Note containerd and hcsshim for HCS v2 APIs do not yet support all the required
functionality needed for docker. gMSA and cloning/templating are two obvious omissions. The gaps will be resolved in time - this PR is a stepping stone in migrating Docker on Windows to containerd for those wishing to experiment, not a full switch-over.

# How to use

You will need RS5 (goal is for RS1+ and although RS1 is currently enabled, there's some issues still being worked through, so it may be disabled prior to merge) and to start with something like the following:

## Window 1:
 - `containerd --log-level debug`

## Window 2:
 - `$env:DOCKER_WINDOWS_CONTAINERD_RUNTIME=1`
 - `dockerd --experimental -D --containerd \\.\pipe\containerd-containerd`

## Required Binaries

You will need the following binary from github.com/containerd/containerd in your path. You will need to make sure this is from master currently, not a 1.2.x branch.
 - `containerd.exe`

(Do **not** use `containerd-shim-runhcs-v1.exe` from the containerd repo - if not already, it will be removed very soon. You should use the one from the https://github.com/Microsoft/hcsshim below)

You will need the following binaries from https://github.com/Microsoft/hcsshim in your path, again currently from 'master' due to the in-flux current status:
 - `containerd-shim-runhcs-v1.exe`
 -  `runhcs.exe` (not required but potentially useful)

For LCOW, the following binaries are required:
 - `C:\Program Files\Linux Containers\initrd.img`
 - `C:\Program Files\Linux Containers\kernel`

This is no different to the current requirements. Linuxkit (https://github.com/linuxkit/lcow) is currently far behind https://github.com/Microsoft/opengcs master, so you may need to build your own initrd.img, as well as kernel. A 4.19 based kernel is preferred.

Commit:
Windows: Experimental: Allow containerd for runtime

Signed-off-by: John Howard <jhoward@microsoft.com>

This is the first step in refactoring moby (dockerd) to use containerd on Windows.
Similar to the current model in Linux, this adds the option to enable it for runtime.
It does not switch the graphdriver to containerd snapshotters.

 - Refactors libcontainerd to a series of subpackages so that either a
  "local" containerd (1) or a "remote" (2) containerd can be loaded as opposed
  to conditional compile as "local" for Windows and "remote" for Linux.

 - Updates libcontainerd such that Windows has an option to allow the use of a
   "remote" containerd. Here, it communicates over a named pipe using GRPC.
   This is currently guarded behind the experimental flag, an environment variable,
   and the providing of a pipename to connect to containerd.

 - Infrastructure pieces such as under pkg/system to have helper functions for
   determining whether containerd is being used.

(1) "local" containerd is what the daemon on Windows has used since inception.
It's not really containerd at all - it's simply local invocation of HCS APIs
directly in-process from the daemon through the Microsoft/hcsshim library.

(2) "remote" containerd is what docker on Linux uses for it's runtime. It means
that there is a separate containerd service running, and docker communicates over
GRPC to it.

To try this out, you will need to start with something like the following:

Window 1:
	containerd --log-level debug

Window 2:
	$env:DOCKER_WINDOWS_CONTAINERD=1
	dockerd --experimental -D --containerd \\.\pipe\containerd-containerd

You will need the following binary from github.com/containerd/containerd in your path:
 - containerd.exe

You will need the following binaries from github.com/Microsoft/hcsshim in your path:
 - runhcs.exe
 - containerd-shim-runhcs-v1.exe

For LCOW, it will require and initrd.img and kernel in `C:\Program Files\Linux Containers`.
This is no different to the current requirements. However, you may need updated binaries,
particularly initrd.img built from Microsoft/opengcs as (at the time of writing), Linuxkit
binaries are somewhat out of date.

Note that containerd and hcsshim for HCS v2 APIs do not yet support all the required
functionality needed for docker. This will come in time - this is a baby (although large)
step to migrating Docker on Windows to containerd.

Note that the HCS v2 APIs are only called on RS5+ builds. RS1..RS4 will still use
HCS v1 APIs as the v2 APIs were not fully developed enough on these builds to be usable.
This abstraction is done in HCSShim. (Referring specifically to runtime)

Note the LCOW graphdriver still uses HCS v1 APIs regardless.

Note also that this does not migrate docker to use containerd snapshotters
rather than graphdrivers. This needs to be done in conjunction with Linux also
doing the same switch.

Commit:
Windows: (WCOW) Generate OCI spec that remote runtime can escape

Signed-off-by: John Howard <jhoward@microsoft.com>

Also fixes https://github.com/moby/moby/issues/22874

This commit is a pre-requisite to moving moby/moby on Windows to using
Containerd for its runtime.

The reason for this is that the interface between moby and containerd
for the runtime is an OCI spec which must be unambigious.

It is the responsibility of the runtime (runhcs in the case of
containerd on Windows) to ensure that arguments are escaped prior
to calling into HCS and onwards to the Win32 CreateProcess call.

Previously, the builder was always escaping arguments which has
led to several bugs in moby. Because the local runtime in
libcontainerd had context of whether or not arguments were escaped,
it was possible to hack around in daemon/oci_windows.go with
knowledge of the context of the call (from builder or not).

With a remote runtime, this is not possible as there's rightly
no context of the caller passed across in the OCI spec. Put another
way, as I put above, the OCI spec must be unambigious.

The other previous limitation (which leads to various subtle bugs)
is that moby is coded entirely from a Linux-centric point of view.

Unfortunately, Windows != Linux. Windows CreateProcess uses a
command line, not an array of arguments. And it has very specific
rules about how to escape a command line. Some interesting reading
links about this are:

https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
https://stackoverflow.com/questions/31838469/how-do-i-convert-argv-to-lpcommandline-parameter-of-createprocess
https://docs.microsoft.com/en-us/cpp/cpp/parsing-cpp-command-line-arguments?view=vs-2017

For this reason, the OCI spec has recently been updated to cater
for more natural syntax by including a CommandLine option in
Process.

What does this commit do?

Primary objective is to ensure that the built OCI spec is unambigious.

It changes the builder so that `ArgsEscaped` as commited in a
layer is only controlled by the use of CMD or ENTRYPOINT.

Subsequently, when calling in to create a container from the builder,
if follows a different path to both `docker run` and `docker create`
using the added `ContainerCreateIgnoreImagesArgsEscaped`. This allows
a RUN from the builder to control how to escape in the OCI spec.

It changes the builder so that when shell form is used for RUN,
CMD or ENTRYPOINT, it builds (for WCOW) a more natural command line
using the original as put by the user in the dockerfile, not
the parsed version as a set of args which loses fidelity.
This command line is put into args[0] and `ArgsEscaped` is set
to true for CMD or ENTRYPOINT. A RUN statement does not commit
`ArgsEscaped` to the commited layer regardless or whether shell
or exec form were used.

ddebroy commented:
@andrey-ko PTAL

jhowardmsft commented:
@ddebroy @andrey-ko Waaaaaaay too early and much is in flux.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -864,7 +864,7 @@ definitions:
         $ref: "#/definitions/HealthConfig"
       ArgsEscaped:
-        description: "Command is already escaped (Windows only)"
+        description: "Command is already escaped (Windows only). Not used and always ignored."
```
I know; too early for review 🤗 - I guess we should "deprecate" this in API v1.41 (older versions of the API, at least when running on an older daemon) would still need it

Commit:
LCOW:Reworking spec builder

Signed-off-by: John Howard <jhoward@microsoft.com>

Commit:
Windows: Fix handle leaks/logging if init proc start fails

Signed-off-by: John Howard <jhoward@microsoft.com>

Fixes #38719

Fixes some subtle bugs on Windows

 - Fixes https://github.com/moby/moby/issues/38719. This one is the most important
   as failure to start the init process in a Windows container will cause leaked
   handles. (ie where the `ctr.hcsContainer.CreateProcess(...)` call fails).
   The solution to the leak is to split out the `reapContainer` part of `reapProcess`
   into a separate function. This ensures HCS resources are cleaned up correctly and
   not leaked.

 - Ensuring the reapProcess goroutine is started immediately the process
   is actually started, so we don't leak in the case of failures such as
   from `newIOFromProcess` or `attachStdio`

 - libcontainerd on Windows (local, not containerd) was not sending the EventCreate
   back to the monitor on Windows. Just LCOW. This was just an oversight from
   refactoring a couple of years ago by Mikael as far as I can tell. Technically
   not needed for functionality except for the logging being missing, but is correct.

cpuguy83 commented:
😜hard to _completely_ ignore with all caps and title makes it intriguing.

jhowardmsft commented:
@cpuguy83 Another week or two :)

Commit:
Windows: Don't shadow err variable

Signed-off-by: John Howard <jhoward@microsoft.com>

Commit:
Vendor Microsoft/hcsshim@ada9cb39

Signed-off-by: John Howard <jhoward@microsoft.com>

jhowardmsft commented:
Hmmm. Hitting https://github.com/containerd/containerd/issues/3023 on all variations after latest vendoring.

Commit:
Windows:Add ETW logging hook

Signed-off-by: John Howard <jhoward@microsoft.com>

Commit:
Vendor sirupsen/logrus@v1.3.0

Signed-off-by: John Howard <jhoward@microsoft.com>

jhowardmsft commented:
@thaJeztah @dmcgowan @crosbymichael @cpuguy83 @johnstep @tiborvass There's still a way to go on this to be what I would call "complete for experimental usage, normal caveats apply, some things may/are known to not work". I'm not pretending by any means it's ready for detailed review as I'm updating/adding commits daily, but some early (and kind haha) feedback as to approach so that we can have a hope of hitting the 1903 cutoff without too much rat-holing with reviews coming in very hot would be appreciated. I suspect the builder/command line change for escaping is probably the most likely part to be heavily scruitinsed and open some 'interesting' discussions 😇. The rest is (IMO) fairly innocuous.

thaJeztah commented:
Thanks @jhowardmsft - added this to the maintainers-review-session (perhaps won't be reviewing it, but to raise awareness of this work, and get early feedback) 👍 🎉

Commit:
Vendor Microsoft/go-winio@4de24ed3

Signed-off-by: John Howard <jhoward@microsoft.com>

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -113,6 +113,7 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 	}
 
 	system.InitLCOW(cli.Config.Experimental)
+	system.InitContainerdOnWindows(cli.Config.Experimental, cli.Config.ContainerdAddr)
```
Does this have to be executed at this point, or could it be at the same point where Linux does this initialisation? Initialisation of systemd on Linux is some lines lower (`if cli.Config.ContainerdAddr == "" && runtime.GOOS != "windows"`).

Perhaps extract to a function, and have different implementations in a `_windows.go` and `_linux.go` file?

jhowardmsft replied:
Yes, that's addressed now.

thaJeztah commented on cmd/dockerd/docker.go:
```diff
@@ -78,6 +78,8 @@ func main() {
 		logrus.SetOutput(stderr)
 	}
 
+	initETWLogging()
```
Given that there's already differences between Windows and Linux above; would it make sense to create a `initLogging` (e.g.) function, and move all Linux/Windows specific initialisation to implementations in a `_windows.go` and `_linux.go` respectively?

jhowardmsft replied:
Yes, that makes sense. Fixed in an update about to be pushed.

thaJeztah commented on daemon/create.go:
```diff
@@ -79,7 +85,7 @@ func (daemon *Daemon) containerCreate(params types.ContainerCreateConfig, manage
 
 // Create creates a new container from the given configuration with a given name.
-func (daemon *Daemon) create(params types.ContainerCreateConfig, managed bool) (retC *container.Container, retErr error) {
+func (daemon *Daemon) create(params types.ContainerCreateConfig, managed bool, ignoreImagesArgEscaped bool) (retC *container.Container, retErr error) {
```
Not a huge fan of the extra boolean argument (the `managed` was already a bit so-so). Perhaps it's time to introduce a `createOpts` struct (or functional arguments)

`daemon. CreateManagedContainer` and `daemon.ContainerCreate` could be moved to `_windows.go` and `_linux.go` files to pass different options / functional arguments

jhowardmsft replied:
I _knew_ you'd say that 😝. I really didn't want to tackle that, but I guess I know what I'll be doing after lunch today....

jhowardmsft replied:
Fixed in latest update.

thaJeztah commented on daemon/errors.go:
```diff
@@ -133,7 +133,8 @@ func translateContainerdStartErr(cmd string, setExitCode func(int), err error) e
 			contains(errDesc, "no such file or directory") ||
-			contains(errDesc, "system cannot find the file specified")) {
+			contains(errDesc, "system cannot find the file specified") ||
+			contains(errDesc, "failed to run runc create/exec call")) {
```
Is this error only created because of an invalid _container_ process? (Couldn't find where this error is generated after a quick search; just want to be sure we're not masking possible issues on the host)

jhowardmsft replied:
Ah, it's returned in the LCOW case from opengcs: https://github.com/Microsoft/opengcs/blob/master/service/gcs/runtime/runc/runc.go#L580

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -326,12 +336,20 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
 // TODO: @jhowardmsft LCOW Support. We need to do a lot more pulling in what can
 // be pulled in from oci_linux.go.
 func (daemon *Daemon) createSpecLinuxFields(c *container.Container, s *specs.Spec) error {
+	s.Root = &specs.Root{
```
(just curious); would, at some point, LCOW be able to just take the "linux" spec, and do the translation of LCOW-specific options in `runhcs` ("ignore" / "warn" about options that aren't supported, and/or translate, e.g. `memory` -> hyper-v memory etc.?)

jhowardmsft replied:
In terms of a spec, generally yes. Using containerd for LCOW outside of moby does just that - it's a Linux spec run on a Linux utility VM so pretty much everything works (or will, at least - still some bugs being worked on in opengcs and containerd).

However, in terms of using the existing Linux code in moby to create a Linux spec, generally no. The current Linux code is too highly dependent on assuming the host is Linux, and much of the processing it does just wouldn't work on Windows.

thaJeztah commented on libcontainerd/remote/client_unix.go:
```diff
@@ -1,4 +1,6 @@
-package libcontainerd // import "github.com/docker/docker/libcontainerd"
+// +build !windows
```
should this file be named `client_linux.go` instead? (it previously was linux-only, now it's "unix", so also built on, e.g. bsd)

jhowardmsft replied:
Oh. I hadn't noticed I switched `_linux.go` to `_unix.go`, wasn't intentional. Good catch. I'll revert that.

jhowardmsft replied:
(Out of interest, is BSD still built, or did that effort die out?)

thaJeztah replied:
I think some people still maintain a fork, and sometimes I've seen a PR come by.

Definitely not a priority, but if we can prevent breaking them, that's always a win :-)

Commit:
Publish empty stats on error

Signed-off-by: John Howard <jhoward@microsoft.com>

jhowardmsft commented on daemon/oci_windows.go:
```diff
@@ -124,15 +123,11 @@ func (daemon *Daemon) createSpec(c *container.Container) (*specs.Spec, error) {
-
 	s.Process.Cwd = c.Config.WorkingDir
 	s.Process.Env = c.CreateDaemonEnvironment(c.Config.Tty, linkedEnv)
+	s.Process.Terminal = c.Config.Tty
```
Puts it logically together with other `s.Process` properties.

jhowardmsft commented on daemon/oci_windows.go:
```diff
@@ -219,11 +221,16 @@ func (daemon *Daemon) createSpec(c *container.Container) (*specs.Spec, error) {
 		return nil, fmt.Errorf("Unsupported platform %q", img.OS)
 	}
 
+	if b, err := json.Marshal(&s); err == nil {
```
Yes, I want this in the final version. It has proved it's worth far too many times now. And yes, I have done the update suggested which makes complete sense. Will push shortly.

jhowardmsft replied:
Also switched to use `logrus.IsLevelEnabled(logrus.DebugLevel)` as this PR also bumps logrus to v1.3.0 (needed for ETW)

jhowardmsft commented on daemon/start_windows.go:
```diff
@@ -1,11 +1,24 @@
 func (daemon *Daemon) getLibcontainerdCreateOptions(container *container.Container) (interface{}, error) {
+
+	// Make sure we set the runhcs options to debug if we are at debug level.
+	if system.ContainerdSupported() && logrus.GetLevel() == logrus.DebugLevel {
```
Resolved, will push shortly. `IsDebugEnabled()` isn't exported. The exported version in logrus v1.3.0 is `logrus.IsLevelEnabled(logrus.DebugLevel)`, so used that instead.

jhowardmsft commented on builder/builder.go:
```diff
@@ -60,8 +60,8 @@ type ImageBackend interface {
-	// ContainerCreate creates a new Docker container and returns potential warnings
-	ContainerCreate(config types.ContainerCreateConfig) (container.ContainerCreateCreatedBody, error)
+	// ContainerCreateIgnoreImagesArgsEscaped creates a new Docker container and returns potential warnings
+	ContainerCreateIgnoreImagesArgsEscaped(config types.ContainerCreateConfig) (container.ContainerCreateCreatedBody, error)
```
Ah, no it's not temporary. This is the backend interface definition. It's an additional entrypoint into the daemon (see change to `daemon\create.go`). `ContainerCreate` has not been removed, just is not called from the builder now.

jhowardmsft commented on builder/dockerfile/dispatchers.go:
```diff
@@ -434,15 +430,23 @@ func prependEnvOnCmd(buildArgs *BuildArgs, buildArgVars []string, cmd strslice.S
+	// We warn here as Windows shell processing operates differently to Linux.
+	// Linux:   /bin/sh -c "echo hello" world	--> hello
+	// Windows: cmd /s /c "echo hello" world	--> hello world
+	if d.state.operatingSystem == "windows" &&
```
I'm kind of loathed to do this in `resolveCmdLine` meaning it would even more parameters. I think having it in the actual dispatcher function for the command makes more sense and is more obvious.

jhowardmsft commented on builder/dockerfile/dispatchers.go:
```diff
@@ -477,8 +481,22 @@ func dispatchHealthcheck(d dispatchRequest, c *instructions.HealthCheckCommand)
-	cmd := resolveCmdLine(c.ShellDependantCmdLine, runConfig, d.state.operatingSystem)
+	cmd, argsEscaped := resolveCmdLine(c.ShellDependantCmdLine, runConfig, d.state.operatingSystem, c.Name(), c.String())
+
+	// This warning is a little more complex than in dispatchCmd(), as the Windows base images (similar
```
See previous response.

Commit:
Remove unsupported lcow.vhdx option

Signed-off-by: John Howard <jhoward@microsoft.com>

This was only experimental and removed from opengcs. Making same
change in docker.

Commit:
Vendor Microsoft/opengcs@a1096715

Signed-off-by: John Howard <jhoward@microsoft.com>

jhowardmsft commented:
@thajeztah - Any idea what this Janky failure might be? I don't think I've changed anything in this area, but maybe I'm missing something:

```
00:02:25.146 The result of hack/generate-swagger-api.sh differs
00:02:25.146 
00:02:25.146 diff --git a/api/types/container/container_wait.go b/api/types/container/container_wait.go
00:02:25.146 index 9e3910a6b4..06b0f02077 100644
00:02:25.146 --- a/api/types/container/container_wait.go
00:02:25.146 +++ b/api/types/container/container_wait.go
00:02:25.146 @@ -7,14 +7,6 @@ package container
00:02:25.146  // See hack/generate-swagger-api.sh
00:02:25.146  // ----------------------------------------------------------------------------
00:02:25.146  
00:02:25.146 -// ContainerWaitOKBodyError container waiting error, if any
00:02:25.146 -// swagger:model ContainerWaitOKBodyError
00:02:25.146 -type ContainerWaitOKBodyError struct {
00:02:25.146 -
00:02:25.146 -	// Details of an error
00:02:25.146 -	Message string `json:"Message,omitempty"`
00:02:25.146 -}
00:02:25.146 -
00:02:25.146  // ContainerWaitOKBody OK response to ContainerWait operation
00:02:25.146  // swagger:model ContainerWaitOKBody
00:02:25.146  type ContainerWaitOKBody struct {
00:02:25.146 @@ -27,3 +19,11 @@ type ContainerWaitOKBody struct {
00:02:25.146  	// Required: true
00:02:25.146  	StatusCode int64 `json:"StatusCode"`
00:02:25.146  }
00:02:25.146 +
00:02:25.146 +// ContainerWaitOKBodyError container waiting error, if any
00:02:25.146 +// swagger:model ContainerWaitOKBodyError
00:02:25.146 +type ContainerWaitOKBodyError struct {
00:02:25.146 +
00:02:25.146 +	// Details of an error
00:02:25.146 +	Message string `json:"Message,omitempty"`
00:02:25.146 +}
00:02:25.146 
00:02:25.146 Please update api/swagger.yaml with any api changes, then 
00:02:25.146 run `hack/generate-swagger-api.sh`.
00:02:25.444 Build step 'Execute shell' marked build as failure
00:02:25.444 Set build name.
00:02:25.445 New build name is '#38541'
00:02:25.477 [PostBuildScript] - Executing post build scripts.
00:02:25.482 [Docker-PRs] $ /bin/sh -xe /tmp/jenkins2917186211448752333.sh
00:02:25.486 + echo Ensuring container killed.
00:02:25.486 Ensuring container killed.
00:02:25.486 + docker rm -vf docker-pr53153
00:02:25.684 Error: No such container: docker-pr53153
00:02:25.686 + true
00:02:25.686 + echo Chowning /workspace to jenkins user
00:02:25.686 Chowning /workspace to jenkins user
00:02:25.686 + id -u
00:02:25.688 + id -g
00:02:25.689 + docker run --rm -v /home/ubuntu/workspace/Docker-PRs:/workspace busybox chown -R 1000:1000 /workspace
00:02:26.620 [Docker-PRs] $ /bin/sh -xe /tmp/jenkins7205177149784333262.sh
00:02:26.630 + echo Creating bundles.tar.gz
00:02:26.630 Creating bundles.tar.gz
00:02:26.630 + xargs tar -czf bundles.tar.gz
00:02:26.630 + find bundles -name *.log -o -name *.prof -o -name integration.test
00:02:26.630 tar: Cowardly refusing to create an empty archive
00:02:26.630 Try 'tar --help' or 'tar --usage' for more information.
00:02:26.630 + true
00:02:26.638 Archiving artifacts
00:02:26.641 Notifying endpoint with url 'https://leeroy.dockerproject.org/notification/jenkins'
00:02:27.168 [WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
00:02:27.598 Finished: FAILURE
```

jhowardmsft commented:
There's some minor further updates needed, but this is ready for review. Have updated the title to match. Let's try and get this into 1903. Biggest outstanding issue is some networking related stuff being followed up on which impacts performance.

thaJeztah commented:
> Any idea what this Janky failure might be? I don't think I've changed anything in this area, but maybe I'm missing something:

Ah, yes, that's because go-swagger uses a `map[string]` somewhere in it's code, therefore making the ordering random 😞. See https://github.com/moby/moby/pull/36714

cpuguy83 commented on daemon/create.go:
```diff
@@ -21,25 +21,46 @@ import (
+
+// ContainerCreateIgnoreImagesArgsEscaped creates a regular container. This is called from the builder RUN case
+// and ensures that we do not take the images ArgsEscaped
+func (daemon *Daemon) ContainerCreateIgnoreImagesArgsEscaped(params types.ContainerCreateConfig) (containertypes.ContainerCreateCreatedBody, error) {
```
Wary of adding another one of these things.

jhowardmsft replied:
Me too. But entirely unavoidable. I spent close on 3 weeks full-time working out how to do this so the end-to-end scenario works. I don't like it, but necessary. This was the only solution that I found to make it work outside of re-engineering docker to accept universally that Windows works through command lines, Linux works through args.

thaJeztah replied:
Should we make the existing `ContainerCreate` accept a `ops ...` so that Windows can call it and pass an `WithIgnoreArgsEscaped()` option?

cpuguy83 replied:
Really, everything in here is terrible...  I'm fine adding this as is b/c such a change means several other unrelated changes.

I do have a branch where I'm working on rewriting all of this code and moving it all into a container service (like we have an image service and a volume service).

cpuguy83 commented on daemon/oci_utils.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// func (daemon *Daemon) addCustomInit(c *container.Container, s *specs.Spec) error {
```
Remove?

jhowardmsft replied:
Ah, yes. Removed in next push.

cpuguy83 commented on daemon/start_windows.go:
```diff
@@ -21,8 +33,6 @@ func (daemon *Daemon) getLibcontainerdCreateOptions(container *container.Contain
 				config.KernelFile = v
 			case "lcow.initrd":
 				config.InitrdFile = v
-			case "lcow.vhdx":
```
Is this breaking something?

jhowardmsft replied:
It's never been "supported" (bearing in mind lcow has to-date always been experimental from docker calling it).

cpuguy83 commented on libcontainerd/local/local_windows.go:
```diff
@@ -1374,3 +1371,40 @@ func (c *client) reapProcess(ctr *container, p *process) int {
+	if err := ctr.hcsContainer.Close(); err != nil {
+		exitCode = -1
+		logger.WithError(err).Error("failed to clean hcs container resources")
+		thisErr := fmt.Errorf("failed to terminate container: %s", err)
```
`errors.Wrap`

jhowardmsft replied:
Fixed in next push

cpuguy83 commented on libcontainerd/local/local_windows.go:
```diff
@@ -1374,3 +1371,40 @@ func (c *client) reapProcess(ctr *container, p *process) int {
+		logger.WithError(err).Error("failed to clean hcs container resources")
+		thisErr := fmt.Errorf("failed to terminate container: %s", err)
+		if eventErr != nil {
+			eventErr = fmt.Errorf("%s: %s", eventErr, thisErr)
```
errors.Wrap

jhowardmsft replied:
> Are the package moves due to platform dependent code imports?

@cpuguy83 No, it's because we are now in this interim stage where previously, "remote" was only called from linux (due to conditional compile) and "local" was only called from Windows (again due to conditional compile). We're temporarily in a hybrid state where Windows can use both local (the default) and remote (the experimental option). Ultimately local will be removed, remote will be the only thing in libcontainerd, things will be more normal-looking, but it's a little way off yet.

jhowardmsft replied:
> errors.Wrap

Fixed in next push

cpuguy83 reviewed (changes requested):
Are the package moves due to platform dependent code imports?

jhowardmsft commented on libcontainerd/local/local_windows.go:
```diff
@@ -1196,22 +1223,22 @@ func (c *client) getProcess(containerID, processID string) (*container, *process
 	switch {
 	case ctr == nil:
-		return nil, nil, errors.WithStack(newNotFoundError("no such container"))
+		return nil, nil, errors.WithStack(libcontainerdtypes.NewNotFoundError("no such container"))
```
That sounds like a possible follow-up, but I don't think it affects the functionality of this PR itself?

cpuguy83 replied:
Why follow-up? This is adding new exported types that's not really doing anything. I don't feel like this is a particularly large change.

jhowardmsft replied:
Ah, sorry, I misunderstood what you meant. Fixed in next push.

jhowardmsft commented on cmd/dockerd/daemon_unix.go:
```diff
@@ -145,3 +148,35 @@ func newCgroupParent(config *config.Config) string {
 	return cgroupParent
 }
+
+func (cli *DaemonCli) initContainerD() (context.Context, context.CancelFunc, error) {
```
Fixed in next push.

jhowardmsft commented:
Have pushed a bunch of feedback related commits. Will squash into the relevant commit once CI passes. @cpuguy83 I think that's all your feedback addressed.

thaJeztah commented:
Looks like there's a build failure;

```
20:14:02 cmd/dockerd/daemon_unix.go:159:31:warning: undeclared name: ctx (gosimple)
20:14:02 cmd/dockerd/daemon_unix.go:159:31:warning: undeclared name: ctx (interfacer)
20:14:02 cmd/dockerd/daemon_unix.go:159:31:warning: undeclared name: ctx (unconvert)
```

jhowardmsft commented:
> Looks like there's a build failure;

Thanks @thaJeztah Ugh, yes, introduced when addressing feedback, I'll fix that tonight, is a simple update. Only affects the Linux side which I didn't compile locally.

jhowardmsft commented:
Green again :)

jhowardmsft commented:
@johnstep - Do you have time to look at this?

cpuguy83 commented on libcontainerd/remote/client_linux.go:
```diff
@@ -1,4 +1,6 @@
-package libcontainerd // import "github.com/docker/docker/libcontainerd"
+// +build !windows
```
Should already be linux only.

jhowardmsft replied:
Fixed.

cpuguy83 reviewed (approved):
This seems relatively straight forward, there are only a couple of places that are a bit 😞 , but that's because the interfaces were already bad to begin with, and the changes aren't making them worse per-se, just more loaded.

Overall, LGTM.
Is CI actually setup to run with the containerd runtime on Windows?

jhowardmsft commented:
> Is CI actually setup to run with the containerd runtime on Windows?

@cpuguy83 No, that's actually quite a large chunk of work and has many hurdles to overcome (not the least of which is binary distribution).   And thanks for reviewing 👍

Commit:
Windows: Implement docker top for containerd

Signed-off-by: John Howard <jhoward@microsoft.com>

Commit:
LCOWv1:Remote lcow.kernel and lcow.initrd

Signed-off-by: John Howard <jhoward@microsoft.com>

LCOWv1 will be deprecated soon anyway (and LCOW is experimental regardless).
Removing lcow.initrd and lcow.kernel options which will not be supported
in LCOWv2 (via containerd).

andrewhsu commented:
I'll see if there is something that can be done to modify the PR jobs to be able to run the tests.

jhowardmsft commented:
@andrewhsu Cool. I value any help. It's a huge job to undertake. But please bear in mind, I can't accept anything that doesn't take `Invoke-DockerCI` into the equation - that's used internally on nightly Windows builds, and can't be broken. As well as by Windows devs in general.

I would suggest that an alternate starting point, while not directly related, would be looking into how to get `integration` (not `integration-cli`) running on Windows CI. Some groundwork has been done between myself and @vdemeester carrying what I started. Both in `hack\windows.ps1` and `Invoke-DockerCI`. We really need that coverage as `integration-cli` is locked for additions. That allows us an in-road into getting LCOW and xenon coverage, all of which are desperately needed in containerd-land. 

Obviously the binary .exe distribution issues (containerd, the runhcs shim, linux kernel, initrd.img are orthogonal, but equally need resolving for the end-to-end)

#ThisIsAHugePieceOfWork #dontunderestimatehowmuchwork #really-iamnotkiddinghere #Maybe1909ReleaseForOnlyContainerDInDockerOnWindowsWithNoInprocHCSInvocation

jhowardmsft commented:
@crosbymichael @dmcgowan Guys - this was discussed again at the maintainers meeting today. We really, really want to make sure this hits the 1903 cut-off. Any chance you guys could review soon? (Obviously appreciate you're busy on other things.)

AkihiroSuda commented on pkg/system/init_windows.go:
```diff
@@ -10,3 +22,19 @@ func InitLCOW(experimental bool) {
+// also requires an environment variable to be set (so as not to turn the
+// feature on from simply experimental which would also mean LCOW.
+func InitContainerdRuntime(experimental bool, cdPath string) {
+	if experimental && len(cdPath) > 0 && len(os.Getenv("DOCKER_WINDOWS_CONTAINERD_RUNTIME")) > 0 {
```
nit: `DOCKER_WINDOWS_CONTAINERD_RUNTIME=0` should disable containerd?

jhowardmsft replied:
@akihirosuda It's its presence rather than it's value.

thaJeztah commented on builder/dockerfile/dispatchers.go:
```diff
@@ -434,15 +430,23 @@ func prependEnvOnCmd(buildArgs *BuildArgs, buildArgVars []string, cmd strslice.S
+	if d.state.operatingSystem == "windows" &&
+		len(runConfig.Entrypoint) > 0 &&
+		d.state.runConfig.ArgsEscaped != argsEscaped {
+		fmt.Fprintf(d.builder.Stderr, " ---> [Warning] Shell-form ENTRYPOINT and exec-form CMD may have unexpected results\n")
```
Do we have a similar warning in place already on the BuildKit side? (just so that we don't forget)

jhowardmsft replied:
I’m not even sure buildkit works on Windows (never tried). It will need more than these warnings as it would need to mirror what this PR does and move to a command line.

thaJeztah reviewed (approved):
LGTM

As a follow-up, could you have a look at;

Adding the containerd and runhcs versions to `docker version` (filled using `fillPlatformVersion`, which currently is a no-op on Windows:

https://github.com/moby/moby/blob/9c83848fc95a200c0c69a34cdeeb1f18a090c13a/daemon/info_unix.go#L119-L128

Add something to the `docker info` output to show that this feature is enabled (similar to how `LCOW` is shown?)

thaJeztah commented:
I think this should be ready to merge; I can wait for SF to wake up for an "ack", but otherwise, ready-to-go (thanks so much for the work on this!!)

We'll also need some documentation; probably in https://github.com/docker/cli/tree/master/experimental

jhowardmsft commented:
@thajeztah. Yes I can do those follow ups. Might need a little more work as we don’t use runhcs in the unified shim model in containerd, but I’ll figure something out.

thaJeztah commented:
Ah, well, whatever information would be useful then (instead of runhcs) 😅

andrewhsu commented:
@andrey-ko when you get a chance, can you have a look a verifying this PR with windows binaries according to the john howard instructions? cc @ddebroy

andrewhsu commented:
@jhowardmsft in regards to getting PR jobs modified to test with containerd on windows in the future and getting nightly builds up with the new dependent binaries, we should proceed with that in parallel with this PR cc @dave-tucker 

(could use msft help on this if you guys got canonical ways to build windows binaries from golang in an automated pipeline)

StefanScherer commented:
The `runhcs.exe` is required, I otherwise got this I couldn't run or kill containers.

```
PS C:\Program Files\Docker> docker run mcr.microsoft.com/windows/nanoserver:1809 ipconfig
...
C:\Program Files\Docker\docker.exe: Error response from daemon: exec: "runhcs.exe": executable file not found in %!P(MISSING)ATH%!(NOVERB): unknown.
```

jhowardmsft commented:
Interesting. It shouldn’t be needed. I’ll fix it as a followup

thaJeztah commented:
```
%!P(MISSING)ATH%!(NOVERB)
```

Looks like an incorrect format-string somewhere as well 🤔

jhowardmsft commented:
@StefanScherer doesn't repro here. Are you POSITIVE you are using the right shim binary?

StefanScherer commented:
@jhowardmsft I tried a Dockerfile using double quotes.

```Dockerfile
# escape=`
ARG core=mcr.microsoft.com/windows/servercore:1809
ARG target=mcr.microsoft.com/windows/servercore:1809
FROM $core as download

SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]

ENV GPG_VERSION 2.3.4

RUN Invoke-WebRequest "https://files.gpg4win.org/gpg4win-vanilla-${env:GPG_VERSION}.exe" -OutFile "gpg4win.exe" -UseBasicParsing ; `
    Start-Process .\gpg4win.exe -ArgumentList '/S' -NoNewWindow -Wait
```

Single line RUN instructions seem to work 🎉 💯. What a releaf.
But I have problems with this multi-line

```Dockerfile
RUN @( 
    "94AE36675C464D64BAFA68DD7434390BDBE9B9C5", 
    "FD3A5288F042B6850C66B31F09FE44734EB7990E", 
    "71DCFD284A79C3B38668286BC97EC7A07EDE3FC1", 
    "DD8F2338BAE7501E3DD5AC78C273792F7D83545D", 
    "C4F0DFFF4E8C1A8236409D08E73BC641CC11F4C8", 
    "B9AE9905FFD7803F25714661B63B535A4C206CA9", 
    "77984A986EBC2AA786BC0F66B01FBB92821C587A", 
    "8FCCA13FEF1D0C2E91008E09770F7A9A5AE15600", 
    "4ED778F539E3634C779C87C6D7062848A1AB005C", 
    "A48C2BEE680E841632CD4E44F07496B3EB3C1762", 
    "B9E2F5981AA6E0CD28160D9FF13993A75599653C" 
    ) | foreach { 
      gpg --keyserver ha.pool.sks-keyservers.net --recv-keys $_ ; 
    }
```

Sure I can keep single quotes here, but wanted to try out where people might use it.

<img width="1068" alt="Screen Shot 2019-03-19 at 12 33 30 PM" src="https://user-images.githubusercontent.com/207759/54636350-83711000-4a43-11e9-98a7-aade454a2d6c.png">

jhowardmsft commented:
Is that a regression (the multi-line)?

StefanScherer commented:
@jhowardmsft oh the shim was my fault, I accidentally copied the wrong one :-(

```Dockerfile
FROM mcr.microsoft.com/windows/servercore:1809
ENV VERSION=1.2.3
RUN powershell -Command Write-Output "The version is $env:VERSION"
SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
RUN Write-Output "The version is $env:VERSION"
```

Building this Dockerfile gives me

```
PS C:\node\quoteetst> docker build --no-cache -t test .
Sending build context to Docker daemon  2.048kB
Step 1/5 : FROM mcr.microsoft.com/windows/servercore:1809
 ---> 1206cbf9524c
Step 2/5 : ENV VERSION=1.2.3
 ---> Running in 1bd568e4ee34
Removing intermediate container 1bd568e4ee34
 ---> 0a74f370c77f
Step 3/5 : RUN powershell -Command Write-Output "The version is $env:VERSION"
 ---> Running in d6a007ab1b1e
The
version
is
1.2.3
Removing intermediate container d6a007ab1b1e
 ---> 0342b5e97e70
Step 4/5 : SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
 ---> Running in 37f64078846b
Removing intermediate container 37f64078846b
 ---> 256f3d5579da
Step 5/5 : RUN Write-Output "The version is $env:VERSION"
 ---> Running in 0b5f8071c718
The
version
is
1.2.3
Removing intermediate container 0b5f8071c718
 ---> a27d43c226db
Successfully built a27d43c226db
Successfully tagged test:latest
```

The output is multi-line so there were no quotes in the Write-Output command. My download from the previous Dockerfile fortunately worked without the quotes.

andrey-ko commented:
I've access problems with container removal: ```

jhowardmsft commented:
Removal issue known and understood. Needs a fix to the shim which is in flight

jhowardmsft commented:
@StefanScherer I’m still trying to understand if this is a regression or not. Can you be more precise?

StefanScherer commented:
@jhowardmsft I think I remembered something wrong. I kept in my mind that I sometime should test this PR to check an enhancement how we can use quotes in RUN instructions. As I finally had some time today to set everything up I tried it out in the hope that we can use double quotes, but it seems I remembered this wrong. So no regression, it just doesn't work as it would be nice to work. So forget about this false alarm. We should track this in another issue.

All other things with containerd work for me on a Server 2019, even docker run --rm.

jhowardmsft commented:
I’m going to merge this so we can move forward given there are three LGTMs and a slack-LGTM from Derek. @thajeztah YOLO.

jhowardmsft commented:
( channeling @jessfraz 😱 😁)

thaJeztah commented:
😂🥳🎉
